src/zigbee2mqtt/zigbee2mqtt-property.ts

```diff
@@ -0,0 +1,246 @@
+  }
+}
+
+export class OnOffProperty extends Zigbee2MqttProperty<boolean> {
+  constructor(
+    device: Zigbee2MqttDevice,
+    name: string,
+    expose: Expos,
+    client: mqtt.Client,
+    deviceTopic: string,
+    debug: boolean
+  ) {
+    super(device, name, expose, client, deviceTopic, debug, {
+      '@type': 'OnOffProperty',
+      title: 'On',
+      type: parseType(expose.type),
+    });
+  }
+
+  update(value: string, update: Record<string, unknown>): void {
+    super.update(value === 'ON', update);
+  }
+
+  protected async sendValue(value: boolean): Promise<void> {
+    return super.sendValue(value ? 'ON' : 'OFF');
+  }
+}
+
+export class BrightnessProperty extends Zigbee2MqttProperty<number> {
+  constructor(
+    device: Zigbee2MqttDevice,
+    name: string,
+    private expose: Expos,
+    client: mqtt.Client,
+    deviceTopic: string,
+    debug: boolean
+  ) {
+    super(device, name, expose, client, deviceTopic, debug, {
+      '@type': 'BrightnessProperty',
+      title: 'Brightness',
+      minimum: 0,
+      maximum: 100,
+      type: 'number',
+    });
+  }
+
+  update(value: number, update: Record<string, unknown>): void {
+    const percent = Math.round((value / (this.expose.value_max ?? 100)) * 100);
+    super.update(percent, update);
+  }
+
+  protected async sendValue(value: number): Promise<void> {
+    return super.sendValue(Math.round((value / 100) * (this.expose.value_max ?? 100)));
+  }
+}
+
+function miredToKelvin(mired: number): number {
+  return Math.round(1_000_000 / mired);
+}
+
+function kelvinToMiredd(kelvin: number): number {
+  return Math.round(1_000_000 / kelvin);
+}
+
+export class ColorTemperatureProperty extends Zigbee2MqttProperty<number> {
+  constructor(
+    device: Zigbee2MqttDevice,
+    name: string,
+    expose: Expos,
+    client: mqtt.Client,
+    deviceTopic: string,
+    debug: boolean
+  ) {
+    super(device, name, expose, client, deviceTopic, debug, {
+      '@type': 'ColorTemperatureProperty',
+      title: 'Color temperature',
+      type: 'number',
+      minimum: miredToKelvin(expose.value_max!),
+      maximum: miredToKelvin(expose.value_min!),
+      unit: 'kelvin',
+    });
+  }
+
+  update(value: number, update: Record<string, unknown>): void {
+    super.update(miredToKelvin(value), update);
+  }
+
+  protected async sendValue(value: number): Promise<void> {
+    return super.sendValue(kelvinToMiredd(value));
+  }
+}
+
+export class ColorProperty extends Zigbee2MqttProperty<string> {
+  constructor(
+    device: Zigbee2MqttDevice,
+    name: string,
+    expose: Expos,
+    client: mqtt.Client,
+    deviceTopic: string,
+    debug: boolean
+  ) {
+    super(device, name, expose, client, deviceTopic, debug, {
+      '@type': 'ColorProperty',
+      title: 'Color',
+      type: 'string',
+      readOnly: false,
+    });
+  }
+
+  update(value: XY, update: Record<string, unknown>): void {
+    const rgb = xyBriToRgb(value.x, value.y, (update.brightness as number) ?? 255);
```

Comment:
I figured out how to do this with the `color` module that we already have:

```typescript
import Color from 'color';

super.update(Color({
  x: value.x * 100,
  y: value.y * 100,
  z: ((update.brightness as number) ?? 255) * 100 / 255,
}).hex(), update);
```

Reply:
There seems to be an error in the calculation logic of the color class.
I added a comment on why the custom function is necessary.